packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts

```diff
@@ -0,0 +1,468 @@
+import { Construct } from 'constructs';
+import { CfnGatewayRoute, CfnVirtualGateway } from './appmesh.generated';
+import { GatewayRoute, GatewayRouteBaseProps } from './gateway-route';
+
+import { IMesh, Mesh } from './mesh';
+import { AccessLog, HealthCheck, PortMapping, Protocol } from './shared-interfaces';
+import { validateHealthChecks } from './utils';
+
+/**
+ * Interface which all Virtual Gateway based classes must implement
+ */
+export interface IVirtualGateway extends cdk.IResource {
+  /**
+   * Name of the VirtualGateway
+   *
+   * @attribute
+   */
+  readonly virtualGatewayName: string;
+
+  /**
+   * The Amazon Resource Name (ARN) for the VirtualGateway
+   *
+   * @attribute
+   */
+  readonly virtualGatewayArn: string;
+
+  /**
+   * The mesh which the VirtualGateway belongs to
+   */
+  readonly mesh: IMesh;
+
+  /**
+   * Utility method to add a list of listeners to this VirtualGateway
+   */
+  addListeners(listeners: VirtualGatewayListener[]): void;
+
+  /**
+   * Utility method to add a new GatewayRoute to the VirtualGateway
+   */
+  addGatewayRoute(id: string, route: GatewayRouteBaseProps): GatewayRoute;
+}
+
+/**
+ * Represents the properties needed to define HTTP Listeners for a VirtualGateway
+ */
+export interface HttpGatewayListenerProps {
+  /**
+   * Port to listen for connections on
+   *
+   * @default - 8080
+   */
+  readonly port?: number
+
+  /**
+   * The health check information for the listener
+   *
+   * @default - no healthcheck
+   */
+  readonly healthCheck?: HealthCheck;
+}
+
+/**
+ * Represents the properties needed to define GRPC Listeners for a VirtualGateway
+ */
+export interface GrpcGatewayListenerProps {
+  /**
+   * Port to listen for connections on
+   *
+   * @default - 8080
+   */
+  readonly port?: number
+
+  /**
+   * The health check information for the listener
+   *
+   * @default - no healthcheck
+   */
+  readonly healthCheck?: HealthCheck;
+}
+
+/**
+ * Represents the properties needed to define listeners for Virtual Gateways
+ */
+export abstract class VirtualGatewayListener {
+  public static httpGatewayListener(props?: HttpGatewayListenerProps): VirtualGatewayListener {
+    return new HttpGatewayListener(props);
+  }
+
+  public static http2GatewayListener(props?: HttpGatewayListenerProps): VirtualGatewayListener {
+    return new Http2GatewayListener(props);
+  }
+
+  public static grpcGatewayListener(props?: GrpcGatewayListenerProps) {
+    return new GrpcGatewayListener(props);
+  }
+
+  public abstract bind(scope: cdk.Construct): CfnVirtualGateway.VirtualGatewayListenerProperty;
+}
+
+export class HttpGatewayListener extends VirtualGatewayListener {
+  /**
+   * Port to listen for connections on
+   *
+   * @default - 8080
+   */
+  readonly port: number;
+
+  /**
+   * Health checking strategy upstream nodes should use when communicating with the listener
+   *
+   * @default - no healthcheck
+   */
+  readonly healthCheck?: HealthCheck;
```

Comment:
I would like to replace `HealthCheck` with protocol variants as well. e.g. `HttpHealthCheck`. Will tackle this as part of a different review though. This one probably has enough in it :)